server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1098,6 +1102,122 @@ private void asyncCleanupUnlinkedShardLevelBlobs(
         }
     }
 
+    public static void remoteDirectoryCleanupAsync(
+        RemoteSegmentStoreDirectoryFactory remoteDirectoryFactory,
+        ThreadPool threadpool,
+        String remoteStoreRepoForIndex,
+        String indexUUID,
+        ShardId shardId
+    ) {
+        threadpool.executor(ThreadPool.Names.REMOTE_PURGE)
+            .execute(
+                new RemoteStoreShardCleanupTask(
+                    () -> RemoteSegmentStoreDirectory.remoteDirectoryCleanup(
+                        remoteDirectoryFactory,
+                        remoteStoreRepoForIndex,
+                        indexUUID,
+                        shardId
+                    ),
+                    threadpool.executor(ThreadPool.Names.REMOTE_PURGE),
+                    indexUUID,
+                    shardId
+                )
+            );
+    }
+
+    /**
+    A Runnable wrapper to make sure that for a given shard only one cleanup task runs at a time.
+     */
+    public static class RemoteStoreShardCleanupTask implements Runnable {
+        private final Runnable task;
+        private final ExecutorService executor;
+        private final String shardIdentifier;
+        final static Set<String> ongoingRemoteDirectoryCleanups = newConcurrentSet();
+        final static ConcurrentMap<String, BlockingQueue<Runnable>> shardCleanupPendingTasks = new ConcurrentHashMap<>();
+
+        public RemoteStoreShardCleanupTask(Runnable task, ExecutorService executor, String indexUUID, ShardId shardId) {
+            this.task = task;
+            this.shardIdentifier = indexShardIdentifier(indexUUID, shardId);
+            this.executor = executor;
+        }
+
+        private static String indexShardIdentifier(String indexUUID, ShardId shardId) {
+            return String.join("/", indexUUID, String.valueOf(shardId.id()));
+        }
+
+        @Override
+        public void run() {
+            if (ongoingRemoteDirectoryCleanups.add(shardIdentifier)) {
+                try {
+                    task.run();
+                    BlockingQueue<Runnable> pendingTasks = shardCleanupPendingTasks.get(shardIdentifier);
+                    try {
+                        if (pendingTasks != null) {
+                            for (Runnable pendingTask = pendingTasks.poll(0L, TimeUnit.MILLISECONDS); pendingTask != null;) {
+                                pendingTask.run();
+                            }
+                        }
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                } finally {
+                    ongoingRemoteDirectoryCleanups.remove(shardIdentifier);
+                }
+            } else {
+                shardCleanupPendingTasks.putIfAbsent(shardIdentifier, new LinkedBlockingQueue<>());
```

Comment:
I see an issue where the control lands on else flow just after the poll method returns null. This will lead to the newly added task will not be picked up until there is a deletion task enqueued again for the same shard.

Reply:
Let's add more tests for this.

Reply:
> I see an issue where the control lands on else flow just after the poll method returns null. This will lead to the newly added task will not be picked up until there is a deletion task enqueued again for the same shard.

I tried to handle that scenario to most extent with the latest implementation but that is also a best effort at the moment since there can still be a race condition which cannot be handled without introducing lock/semaphores however adding locks/semaphores introduces the overhead of lock object cleanup handling.
Will create a followup task for fixing this race condition. also since the long term plan is to introduce remote store GC component (https://github.com/opensearch-project/OpenSearch/issues/8469), this is just a short term fix.